### PR TITLE
Introducing Session, cleaning UI-facing API, refactor package structure

### PR DESCRIPTION
**Pull request by astaff**
This PR is largely influenced by this proposal: https://docs.google.com/document/d/1fXJBd1SFIqudxdWzyFSHoq5mkER45tr_A5Yx1s-nFQs/edit#heading=h.rkdjwohfjnc0 and feedback from @mcous based on making API calls from the UI.

# Session and Session Manager

In order to further separate robot API that is immediately related to the robot, instruments, containers and modules, from actual protocol and running protocols we are introducing a concept of `Session` and `SessionManager`. This concept also allows to run protocols outside python interpreter more intuitively. 

`SessionManager` is responsible for creating and getting an active session. And `Session` is responsible for representing mutable state of a protocol run (execution state, commands, errors, etc). `Session` also contains messages related to running the protocol: `run`, `pause`, etc.

*NOTE:* the PR also re-orders argument sequence for `SessionManager.create`: `name` goes first, `text` goes next.

# Removing intermediary object for making remote API calls

Before this PR, upon establishing a connection to opentrons.server, the client would receive instance and type description of of an intermediary object. This would allow to make system calls such as `get_object_by_id` or `get_root` to get the actual object the server exposes for remote calls. This mechanism introduced unnecessary boilerplate in client and unit tests, requiring to call `get_root` for every new connection before making calls to actual remote object.

This PR removes intermediary and returns root object and it's type upon connecting. To support system calls (`get_object_by_id`), the client should omit `id` key in request message. This will signal the server that the call is being made to an object that wraps system calls.

# Refactor package structure.

* added `opentrons.session` and moved session-related code out of `opentrons.server`
* tests from `opentrons/server/tests` are now in `tests/opentrons/server` and `tests/opentrons/session`

Check List:

- [ ] Are there breaking changes in the API and how are they being communicated?
- [ ] Who is reviewing your code?

### CODE REVIEW

File: api/opentrons/server/rpc.py [hunk 1]
```diff
@@ -6,12 +6,16 @@
...
+from aiohttp import WSCloseCode
 from asyncio import Queue
 from opentrons.server import serialize
-from threading import Thread
+from concurrent.futures import ThreadPoolExecutor
 
 log = logging.getLogger(__name__)
 
+# Number of executor threads
+MAX_WORKERS = 2
```
Comment (mcous) [inline api/opentrons/server/rpc.py]: Question so I can learn (thanks!): does Python have any convention (official or otherwise) for storing configuration stuff like this? e.g. Scala/Akka projects tend to use Typesafe Config, node projects can put config on package.json, dotenv, or any number of user-land config libraries, etc.

File: api/opentrons/server/rpc.py [hunk 2]
```diff
@@ -20,71 +24,24 @@
...
-                asyncio.set_event_loop(loop)
-                loop.run_forever()
-            finally:
-                log.info('Exiting from exec thread')
-                loop.close()
+    def __init__(self, root=None, loop=None, notification_max_depth=5):
+        self.monitor_events_task = None
+        self.loop = loop or asyncio.get_event_loop()
 
         self.objects = {id(self): self}
```
Comment (mcous) [inline api/opentrons/server/rpc.py]: It looks like there might be some overlap between `self.objects` and `self.control_box`. Are both needed? Also, same question here as with `ControlBox`, should the server be storing a reference to itself? Presumably this would allow the server to return itself, which I assume we want to avoid.

File: api/opentrons/server/rpc.py [hunk 3]
```diff
@@ -97,98 +54,85 @@ def root(self):
...
-            try:
-                await self.process(msg)
-            except Exception as e:
-                await ws.close()
-                log.warning(
-                    'While processing message: {0}'
-                    .format(traceback.format_exc()))
+        self.send({
+            '$': {'type': CONTROL_MESSAGE},
+            'root': self.call(lambda: self.root),
```
Comment (mcous) [inline api/opentrons/server/rpc.py]: 👍

File: api/opentrons/session/session.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        except Exception as e:
+            self.set_state('error')
+            _, _, traceback = sys.exc_info()
+            self.error_append((e, traceback))
+        finally:
+            robot.disconnect()
+
+        return (self, robot.commands())
+
+    def set_state(self, state):
```
Comment (mcous) [inline api/opentrons/session/session.py]: Should `set_state` send out some sort of notification?

File: api/opentrons/server/serialize.py [hunk 5]
```diff
@@ -50,7 +47,7 @@ def iterate(kv): return {str(k): object_tree(v) for k, v in kv.items()}
             items = [object_tree(o) for o in obj]
         except TypeError:
             pass
-        tail = {str(i): v for i, v in enumerate(items)}
+        tail = {i: v for i, v in enumerate(items)}
```
Comment (mcous) [inline api/opentrons/server/serialize.py]: Just curious: what are the implications of this change?

File: api/setup.py [hunk 6]
```diff
@@ -30,9 +30,8 @@
     "writing automated biology lab protocols easy.")
 PACKAGES = find_packages(where='.', exclude=["tests.*", "tests"])
 INSTALL_REQUIRES = [
-    'dill==0.2.5',
-    'requests==2.14.2',
-    'pyserial==3.2.1']
+    'pyserial==3.2.1',
+    'aiohttp==2.2.3']
```
Comment (mcous) [inline api/setup.py]: Why are these here and not in `requirements.txt`?

File: api/opentrons/server/rpc.py (hunk 2 above)
Comment (astaff) [reply api/opentrons/server/rpc.py]: Basically `ControlBox` is intended to prevent server from having to store a reference to self.

File: api/opentrons/server/rpc.py [hunk 7]
```diff
@@ -328,7 +276,14 @@ def send(self, payload):
...
             log.info('Enqueuing {0} for {1}'.format(payload, client))
-            client.send_str(json.dumps(payload))
-            # TODO(artyom, 2017-08-24): finish work on queuing responses
-            # and supporting back-pressure
-            # self.loop.create_task(self.send_queue.put((client, payload)))
+            asyncio.run_coroutine_threadsafe(
+                self.send_queue.put((client, payload)), self.loop)
+
+
+class ControlBox(object):
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/server/rpc.py]: Let's sync up — I removed it and then re-introduced. Let's bounce ideas off of each other to come up with cleaner design

File: api/opentrons/server/rpc.py [hunk 8]
```diff
@@ -328,7 +276,14 @@ def send(self, payload):
...
+    def __init__(self, objects):
+        self.objects = objects
+        objects[id(self)] = self
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/server/rpc.py]: That's control box, not server. The intent is to expose all the methods that are being called by the client with `id` not set. These will be delegated to `ControlBox`. The name might sound confusing. Happy to reconsider.

File: api/opentrons/server/serialize.py (hunk 5 above)
Comment (astaff) [reply api/opentrons/server/serialize.py]: We are letting JSON serializer to handle that. If it is not string serializable it will break anyways. As far as our test scenarios are concerned this doesn't change anything.

File: api/opentrons/session/session.py (hunk 4 above)
Comment (astaff) [reply api/opentrons/session/session.py]: Good idea. I will make it emit something like `{'name': 'state-changed', 'arguments': {'state': '...'}`

File: api/opentrons/server/rpc.py (hunk 1 above)
Comment (astaff) [reply api/opentrons/server/rpc.py]: There's no PEP (Python Enhancement Proposal) that addresses configuration. PEP is an official document defining how things should be done. Common ways are inline dictionary or class, environment variables or (sic!) .ini files. Since this constant is something that affects the way we call things and makes certain scenarios possible (like you can't run a protocol and call `stop` with `MAX_WORKERS` set to `2`), i am not sure if there is a benefit of extracting it into something just yet. And ini files are so Windows 3.1 ...

File: api/opentrons/server/main.py [hunk 9]
```diff
@@ -71,7 +71,10 @@ def parse_command_line(argv):
         print(str(e))
         exit(1)
 
-    server = Server(RobotContainer())
+    # TODO(artyom, 20170828): consider moving class name definition into
```
[the remark this answers is not in the document]
Comment (astaff) [reply api/opentrons/server/main.py]: Will add that once we have another service we will be running

File: api/opentrons/server/rpc.py (hunk 1 above)
Comment (mcous) [reply api/opentrons/server/rpc.py]: 👍 Just curious. Regardless, you're right, constant in the file seems perfect until we have an actual need to configure this

File: api/opentrons/server/rpc.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcous) [reply api/opentrons/server/rpc.py]: Womp womp. Thanks

File: api/opentrons/session/session.py (hunk 4 above)
Comment (mcous) [reply api/opentrons/session/session.py]: Just as a reference (not necessarily a good idea to couple client and server like this) but on the redux side of things we are using [Flux standard actions](https://github.com/acdlite/flux-standard-action) as our action format, which might square nicely with a notification format

File: api/setup.py (hunk 6 above)
Comment (astaff) [reply api/setup.py]: `requirements.txt` is something we need to build and test and `setup.py` is something we need to install and run. `pip install` processes `setup.py` to install dependencies. Python packaging and dependency management is an interesting subject

**Comment by mcous** [comment]
Closing in favor of #330
